src/_posts/2017-08-04-splittable-do-fn.md

```diff
@@ -0,0 +1,449 @@
+---
+layout: post
+title:  "Powerful and modular IO connectors with Splittable DoFn in Apache Beam"
+date:   2017-08-15 00:00:01 -0800
+excerpt_separator: <!--more-->
+categories: blog
+authors:
+  - jkff
+---
+
+One of the most important parts of the Apache Beam ecosystem is its quickly
+growing set of connectors allowing Beam pipelines to read and write data to
+various data storage systems ("IOs"). Currently Beam ships [over 20 IO
+connectors]({{ site.baseurl }}/documentation/io/built-in/) with many more
+in active development. As user demands for IO connectors grew, our work on
+improving the related Beam APIs (in particular, the Source API) produced an
+unexpected result: a generalization of Beam's most basic primitive, `DoFn`.
+
+<!--more-->
+
+## Connectors as mini-pipelines
+
+One of the main reasons for this vibrant IO connector ecosystem is that
+developing a basic IO is relatively straightforward: many connectors are
+implemented simply as mini-pipelines (composite `PTransform`s) made of the basic
+Beam `ParDo` and `GroupByKey` primitives.
+
+For example, `ElasticsearchIO.write()`
+[expands](https://github.com/apache/beam/blob/f7e8f886c91ea9d0b51e00331eeb4484e2f6e000/sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java#L783)
+into a single `ParDo` with some batching for performance; `JdcbIO.read()`
+[expands](https://github.com/apache/beam/blob/f7e8f886c91ea9d0b51e00331eeb4484e2f6e000/sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java#L329)
+into `Create.of(query)`, a reshuffle to [prevent
```

Comment:
Some of these descriptions of mini-pipelines might be nice to illustrate. They are all straight line so it won't necessarily be interesting. Up to you.